napari/layers/_layer_actions.py

```diff
@@ -143,10 +158,17 @@ def _project(ll: LayerList, axis: int = 0, mode='max'):
             )
         )
 
+    axis = dims.order[axis]
     # this is not the desired behavior for coordinate-based layers
     # but the action is currently only enabled for 'image_active and ndim > 2'
     # before opening up to other layer types, this line should be updated.
-    data = (getattr(np, mode)(layer.data, axis=axis, keepdims=False),)
+    data = getattr(np, mode)(layer.data, axis=axis, keepdims=False)
+
+    # In case the last 2 dimensions are these axes, the data needs to be swapped in order to not have the projection
+    # displayed orthogonal to the image data.
+    must_swap = ((0, 1), (2, 0))
```

Comment:
I'm slightly confused why it's these two specifically and why not other non-default orders like `(1, 0)`. Would be great to add more explanation or test cases to help!

Reply:
After trying to write some tests, I think the reason for this specific swap (and other related ones) is fundamentally due to the issue described in: https://github.com/napari/napari/issues/3882

For example, in the case that `dim.order == (2, 0, 1)` and `layer.data.shape == (6, 7, 8)`, I'd naively expect the projected data to have a shape of (7, 8) because axis 2 was projected. However, the current implementation makes the projected data have a shape of (8, 7). That actually accounts for our broken dimensional correspondence, which will produce a layer-specific order of (1, 0) rather than (0, 1) from (2, 0, 1).

So the PR anticipates the broken-ness, so that visualization turns out OK. As I think the projection is only available through the GUI, I don't mind this too much. And we effectively fix the specific issue without needing to solve the general dimensional correspondence problem. But it probably suggests that we should write these as integration tests that also incorporates vispy (e.g. see `test_vispy_image_layer`), so that we don't have to update these tests if/when we fix dimensional correspondence. I'm happy to give that a try if this all makes sense to you?

Alternatively, we can try to fix the bigger underlying dimensional correspondence issue, in which this PR won't need to special-case certain orders and just integrate over the first axis in `dims.order`. But that almost certainly will be a lot more work.

Reply:
Thanks for having a look! Yes exactly, that is also how I understood it. Changing the axis order does not change the axis order of the layer data.

I would say perhaps lets go for option 1 for now since it is quicker and then work on the dimension correspondence problem. Particularly the problem right now is that if by use of the gui you create a max projection it is always the same max projection no matter whether you changed the axis order. The max projection always uses axis 0 of layer.data so I think a quicker fix would be better before actually solving the problem. What do you think?

Reply:
I'm going to try a potential fix for the dimension correspondence problem tomorrow and see how much breaks. If it looks bad, I'm supportive of the quick fix here for now with a vispy integration test that will support any future changes.

Reply:
let me know if you want to paircode on that, I will have time in the evening my time:)